### PR TITLE
Gifting

### PR DESCRIPTION
**Pull request by benthecarman**
Used for testing the gifting feature

### CODE REVIEW

**Comment by futurepaul** [comment]
needs #711 merged and then this is ready for review

**Comment by futurepaul** [comment]
things to test:
not enough funds on sender
too small amount for receiver (if it's a fresh wallet)
trying to redeem twice
loading from url
loading from scanner

**Comment by futurepaul** [comment]
rebase achieved, haven't re-tested yet. todo list here: https://github.com/MutinyWallet/mutiny-web/issues/540

**Comment by benalleng** [comment]
When using the qr-scanner I am sent to this page
https://gifting.mutiny-web.pages.dev/gifting.mutiny-web.pages.dev

**Comment by futurepaul** [comment]
> When using the qr-scanner I am sent to this page https://gifting.mutiny-web.pages.dev/gifting.mutiny-web.pages.dev

I think I fixed this, can't test right now though. (it was only a problem because I was looking for the string `/gift` and our staging url includes `/gift` because `//gifting..`

**Comment by TonyGiorgio** [comment]
This warning appears on the amount dialog for the gift amount editor. You should probably default to these types of warnings not being displayed so we quit running into it for anything besides the receive flow.

Screenshot from 2023-09-23 13-26-49

**Comment by TonyGiorgio** [comment]
1. The lightning fee warning is appearing but it wasn't needed. I redeemed 1 gift for 10k sats which opened up my first channel + 100k sats. Then I went to redeem a gift of 50k sats, which it warned I needed another channel for. This was incorrect.
2. Could the timeout warning be better here? In this case, I closed the browser on the gifter side but then redeemed on the receiver side. It should say that the gift is in progress becauase it sent the NWC message and it will be sent whenever the gifter comes back online. 
3. There's nothing on the gifter side warning them that they need to be online for the gift to be redeemed.

Screenshot from 2023-09-23 13-32-17

**Comment by TonyGiorgio** [comment]
Last thing, love the new prompt for mutiny+

Screenshot from 2023-09-23 13-36-59

Can we get some sort of prompt on the home screen too if they haven't subscribed yet? We can probably do something similar to some other things we do by just putting a timestamp when the message was dismissed. Should probably not be present if the backup warning is already there. 

For now we can just make it so that when the user dimissed it, the timestamp is recorded and then we just leave it off. In the future maybe we can reprompt them but maybe later when we have more features.

**Comment by TonyGiorgio** [comment]
In general it's feeling pretty solid! Just a few edge cases and smaller feature requests. Nice work!

**Comment by benalleng** [comment]
Tested on firefox mobile web
- Not enough funds on sender error only happens on the receiver side (should probably check the available lightning balance at the time of gift creation)
- below 50k sats shows an appropriate warning on sender and an error on receiver side
- scanner and url both work as expected

**Comment by TonyGiorgio** [comment]
Should this feature be available to anyone that is self hosting?

File: src/routes/settings/Gift.tsx [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+
+export function SingleGift(props: {
+    profile: NwcProfile;
+    onDelete?: () => void;
+}) {
+    const i18n = useI18n();
+    const [state, _actions] = useMegaStore();
+
+    const baseUrl = window.location.origin;
+    const sharableUrl = () => baseUrl.concat(props.profile.url_suffix || "");
```
[the remark this answers is not in the document]
Comment (TonyGiorgio) [reply src/routes/settings/Gift.tsx:59]: Ah, nevermind it's coming from `window.location.origin` , yeah that won't work. It should default to app.mutinywallet.com if it's native mainnet and signet-app.mutinywallet.com if it's native signet.

**Comment by TonyGiorgio** [comment]
I hardcoded URL on both the sending and receiving side and as long as these match up properly, android works great. Even opens in app and/or scans in app properly. So just detecting android and switching URL should work.

**Comment by benthecarman** [comment]
> Should this feature be available to anyone that is self hosting?

yeah it should

File: src/routes/settings/Gift.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (futurepaul) [reply src/routes/settings/Gift.tsx:59]: how would we test changes to redeem?

[the remark this answers is not in the document]
Comment (TonyGiorgio) [reply src/routes/settings/Gift.tsx:59]: This is on android, there's no local host to run.

[the remark this answers is not in the document]
Comment (futurepaul) [reply src/routes/settings/Gift.tsx:59]: I guess just the inverse of how you just did it

**Comment by futurepaul** [comment]
okay I think warnings are well behaved now and have a better error for timeout

Screenshot 2023-10-03 at 12 57 39 PM

Screenshot 2023-10-03 at 1 55 30 PM

also hardcoded urls

only comment I didn't get to is this one, didn't feel like it's 100% necessary

> Not enough funds on sender error only happens on the receiver side (should probably check the available lightning balance at the time of gift creation)

File: src/routes/settings/Gift.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (TonyGiorgio) [reply src/routes/settings/Gift.tsx:65]: This only needs to be hard coded when android is the native platform, there's capacitor APIs for this.

**Review by TonyGiorgio — CHANGES_REQUESTED** [review]

File: src/routes/settings/Gift.tsx [hunk 2]
```diff
@@ -198,12 +204,15 @@ export default function GiftPage() {
         return Number(getValue(giftForm, "amount")) < 50000;
     };
 
+    const selfHosted = state.mutiny_wallet?.get_network() === "signet";
```
Comment (benthecarman) [inline src/routes/settings/Gift.tsx:207]: something is wrong here, either should check self hosted another way or rename this variable

Comment (futurepaul) [reply src/routes/settings/Gift.tsx:207]: nice catch bad copy pasting by me

**Review by benthecarman — COMMENTED** [review]
ACK

**Review by TonyGiorgio — APPROVED** [review]

**Pull request merged**